### PR TITLE
Separate all jsDependencies-related things in separate projects.

### PR DESCRIPTION
**Pull request by sjrd**
This is part of #2841.

* The contents of `org.scalajs.core.tools.jsdep._` are extracted in a separate project `jsdependencies-core`.
* All the settings related to `jsDependencies` are extracted in a separate sbt plugin `JSDependenciesPlugin` in `jsdependencies-plugin`.

### CODE REVIEW

**Comment by sjrd** [comment]
Dear Jenkins, retest this please

File: sbt-plugin-test/build.sbt [hunk 1]
```diff
@@ -138,8 +138,13 @@ lazy val multiTest = crossProject.
   jsSettings(baseSettings: _*).
   jsSettings(
     name := "Multi test framework test JS",
+
     // Make FrameworkDetector resilient to other output - #1572
-    jsDependencies in Test += ProvidedJS / "consoleWriter.js",
+    jsExecutionFiles in Test := {
+      val consoleWriter = FileVirtualJSFile(
+          baseDirectory.value / "src/test/resources/consoleWriter.js")
```
Comment (sjrd) [inline sbt-plugin-test/build.sbt]: Piece of self-review: use `(resourceDirectory in Test).value / "consoleWriter.js"` instead.

File: jsdependencies-sbt-plugin/src/main/scala/org/scalajs/jsdependencies/sbtplugin/JSDependenciesPlugin.scala [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+            jsDependencyManifests.value.data.exists(_.requiresDOM))
+      }
+  )
+
+  lazy val compileSettings = configSettings
+
+  lazy val testSettings = Def.settings(
+      configSettings,
+
+      moduleName in packageJSDependencies := moduleName.value + "-test"
```
Comment (sjrd) [inline jsdependencies-sbt-plugin/src/main/scala/org/scalajs/jsdependencies/sbtplugin/JSDependenciesPlugin.scala]: This refactoring highlights that `moduleName in packageMinifiedJSDependencies` does not get its `-test` suffix here.

Comment (gzm0) [inline jsdependencies-sbt-plugin/src/main/scala/org/scalajs/jsdependencies/sbtplugin/JSDependenciesPlugin.scala]: I do not understand this. Shouldn't `moduleName in Test` have the "-test" suffix overall?

File: jsdependencies-sbt-plugin/src/main/scala/org/scalajs/jsdependencies/sbtplugin/JSDependenciesPlugin.scala [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+        "The filter applied to JS dependency manifests before resolution", CSetting)
+
+    val resolvedJSDependencies = TaskKey[Attributed[Seq[ResolvedJSDependency]]]("resolvedJSDependencies",
+        "JS dependencies after resolution.", DTask)
+
+    /** Internal task to calculate whether a project requests the DOM
+     *  (through jsDependencies or requiresDOM) */
+    val scalaJSRequestsDOM = TaskKey[Boolean]("scalaJSRequestsDOM",
+        "Scala.js internal: Whether a project really wants the DOM. " +
+        "Calculated using requiresDOM and jsDependencies", KeyRanks.Invisible)
```
Comment (gzm0) [inline jsdependencies-sbt-plugin/src/main/scala/org/scalajs/jsdependencies/sbtplugin/JSDependenciesPlugin.scala]: This setting is unused. We used to use it to determine the `JSEnv` automatically. I think it can just be dropped (OK to do later, if you want clean deprecation).

File: jsdependencies-sbt-plugin/src/main/scala/org/scalajs/jsdependencies/sbtplugin/JSDependenciesPlugin.scala [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+        "classpath in a single file.", AMinusTask)
+
+    val jsDependencyManifest = TaskKey[File]("jsDependencyManifest",
+        "Writes the JS_DEPENDENCIES file.", DTask)
+
+    val jsDependencyManifests = TaskKey[Attributed[Traversable[JSDependencyManifest]]](
+        "jsDependencyManifests", "All the JS_DEPENDENCIES on the classpath", DTask)
+
+    val requiresDOM = SettingKey[Boolean]("requiresDOM",
+        "Whether this projects needs the DOM. Overrides anything inherited through dependencies.", AMinusSetting)
```
Comment (gzm0) [inline jsdependencies-sbt-plugin/src/main/scala/org/scalajs/jsdependencies/sbtplugin/JSDependenciesPlugin.scala]: This setting is unused (short of `scalaJSRequestsDOM`). Same story about dropping than for `scalaJSRequestsDOM`.

File: jsdependencies-sbt-plugin/src/main/scala/org/scalajs/jsdependencies/sbtplugin/JSDependenciesPlugin.scala (hunk 2 above)
Comment (sjrd) [inline jsdependencies-sbt-plugin/src/main/scala/org/scalajs/jsdependencies/sbtplugin/JSDependenciesPlugin.scala]: `moduleName in Test`, not scoped to any task, is not ours to mess with. And sbt just defaults it `moduleName in Compile`, which therefore has no `-test` suffix.

File: ci/matrix.xml [hunk 5]
```diff
@@ -174,7 +174,8 @@
   <task id="tools-cli-stubs"><![CDATA[
     setJavaVersion $java
     sbt ++$scala tools/package ir/test tools/test cli/package cli/assembly \
-        stubs/package jsEnvsTestSuite/test \
+        stubs/package jsEnvsTestSuite/test phantomJSEnv/test \
```
Comment (sjrd) [inline ci/matrix.xml]: Duh, yes, I wanted to make a separate commit for that. I forgot when I made the commit, I guess.

Comment (sjrd) [inline ci/matrix.xml]: Though now that I've already extracted PhantomJS in the other repo, I think I'll just drop that change.

File: jsdependencies-sbt-plugin/src/main/scala/org/scalajs/jsdependencies/sbtplugin/JSDependenciesPlugin.scala (hunk 2 above)
Comment (gzm0) [inline jsdependencies-sbt-plugin/src/main/scala/org/scalajs/jsdependencies/sbtplugin/JSDependenciesPlugin.scala]: Ok, but why do we not want a `-test` suffix for minified dependencies then? (or is this a bug you simply retain?, if yes, add clarifying comment?)

File: jsdependencies-sbt-plugin/src/main/scala/org/scalajs/jsdependencies/sbtplugin/JSDependenciesPlugin.scala (hunk 3 above)
Comment (sjrd) [inline jsdependencies-sbt-plugin/src/main/scala/org/scalajs/jsdependencies/sbtplugin/JSDependenciesPlugin.scala]: I agree that it should disappear/be deprecated, but not in the commit/PR that imports the whole thing from core ;) It was just as useless in core before the import. The import itself did not make it useless.

Comment (gzm0) [inline jsdependencies-sbt-plugin/src/main/scala/org/scalajs/jsdependencies/sbtplugin/JSDependenciesPlugin.scala]: Agreed.

File: jsdependencies-sbt-plugin/src/main/scala/org/scalajs/jsdependencies/sbtplugin/JSDependenciesPlugin.scala (hunk 2 above)
Comment (sjrd) [inline jsdependencies-sbt-plugin/src/main/scala/org/scalajs/jsdependencies/sbtplugin/JSDependenciesPlugin.scala]: We definitely *do* want it. It's currently a bug. I was actually wondering whether to preserve the bug during the extraction and fix it later, once it is in its own repo; or to fix it during the extraction. WDYT?

Comment (gzm0) [inline jsdependencies-sbt-plugin/src/main/scala/org/scalajs/jsdependencies/sbtplugin/JSDependenciesPlugin.scala]: I would retain the current behavior. Just like with the keys, you are right that we should do one thing at a time.

File: ci/matrix.xml (hunk 5 above)
Comment (gzm0) [inline ci/matrix.xml]: Works for me, too.

File: sbt-plugin-test/build.sbt (hunk 1 above)
Comment (sjrd) [inline sbt-plugin-test/build.sbt]: Done.

File: ci/matrix.xml (hunk 5 above)
Comment (sjrd) [inline ci/matrix.xml]: Done.

**Comment by sjrd** [comment]
Jenkins, retest this please

**Comment by sjrd** [comment]
Dear Jenkins, retest this please

**Comment by sjrd** [comment]
Rebased. Only trivial conflicts, no need to re-review.
